File: app/src/main/java/com/mirash/familiar/view/gesturedetect/tool/GestureEventDetector.java
package com.mirash.familiar.view.gesturedetect.tool;

import android.content.Context;
import android.view.GestureDetector;
import android.view.GestureDetector.SimpleOnGestureListener;
import android.view.MotionEvent;

/**
 * @author Mirash
 */
public class GestureEventDetector implements TouchListener {
    private final GestureDetector gestureDetector;
    private boolean isEnabled;
    private TapDetector tapDetector;

    public GestureEventDetector(Context context) {
        gestureDetector = new GestureDetector(context, new GestureListener());
        isEnabled = true;
    }

    public void setSingleTapListener(SingleTapUpListener tapListener) {
        tapDetector = new TapDetector(tapListener);
    }

    public void setEnabled(boolean enabled) {
        isEnabled = enabled;
    }

    private final class GestureListener extends SimpleOnGestureListener {

        @Override
        public boolean onDown(MotionEvent e) {
            return true;
        }

        @Override
        public boolean onSingleTapUp(MotionEvent e) {
            return tapDetector != null && tapDetector.onSingleTapUp();
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return isEnabled && gestureDetector.onTouchEvent(event);
    }
}

File: app/src/main/java/com/mirash/familiar/view/gesturedetect/tool/GestureEventDetector.kt
package com.mirash.familiar.view.gesturedetect.tool

import android.content.Context
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent

/**
 * @author Mirash
 */
class GestureEventDetector(context: Context) : TouchListener {
    private val gestureDetector: GestureDetector = GestureDetector(context, GestureListener())
    private var tapDetector: TapDetector? = null
    var isEnabled: Boolean = true

    fun setSingleTapListener(tapListener: SingleTapUpListener?) {
        if (tapListener != null) {
            tapDetector = TapDetector(tapListener)
        } else {
            tapDetector = null
        }
    }

    private inner class GestureListener : SimpleOnGestureListener() {
        override fun onDown(e: MotionEvent): Boolean = true

        override fun onSingleTapUp(e: MotionEvent): Boolean = tapDetector?.onSingleTapUp() ?: false
    }

    override fun onTouchEvent(event: MotionEvent): Boolean = isEnabled && gestureDetector.onTouchEvent(event)
}